rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+
+In the longer term, blocks can be developed _within_ emergent standards, discovering related descriptions of data, and having greater confidence that they fit the problems well. Users benefit from an ecosystem with less guesswork, and greater specificity. Embedding applications have greater potential to implement programmatic inference, and have a better specified toolkit to utilize when addressing problems of mapping data.
+
+#### Simple use-cases
+
+Although there are numerous [drawbacks](#drawbacks) that affect the simpler use-cases, there are still large benefits to the proposal. Having better-defined types improves machine comprehension, but also improves human comprehension. A developer trying to manually map their data to the API of a block can better understand what the block needs, and more quickly satisfy its requirements.
+
+A block developer trying to create a block that's specific to their domain for their own application might inadvertently create something that's useful to a wider audience, by participating in the wider ecosystem. The chance of them creating something useful is increased through reusability of descriptions.
+
+## Specific Decisions
+
+There are a number of very specific design decisions that are also encapsulated within this proposal, it's worth capturing some of the noteworthy ones.
+
+### Allowing `oneOf`
+
+Representing the Type System within common structures like a Relation Database is greatly complicated by the inclusion of the boolean operators (`oneOf` is equivalent to an `or` operator). It is required under the constraint of being able to map any existing JSON data, as reasoned above. A simple example of this is trying to represent an object (this case occurs either in an Entity or a Property Type object) that looks like the following:
+
+```json
+{
+  "someProperty": [2.0, "foo"]
+}
+```
+
+To be able to represent this object, a Property Type would need to be created that looks like (a simplified view for brevity):
+
+```json
+{
+  "name": "someProperty",
+  "oneOf": [{ "$ref": "data-type/number" }, { "$ref": "data-type/text" }]
+}
+```
+
+More specifically, it's required to allow `oneOf` with more variants than _just Data Types_ because:
+
+```json
+{
+  "someProperty": [{ "bar": 2.0 }, "foo"]
+}
+```
+
+Requires something like
+
+```json
+{
+  "name": "someProperty",
+  "oneOf": [
+    {
+      "type": "object",
+      "properties": {
+        "bar": { "$ref": "property-type/bar" }
+      }
+    },
+    { "$ref": "data-type/text" }
+  ]
+}
+```
+
+### Including an `Empty List` Data Type
+
+Similarly, the `Empty List` Data Type is required to be able to programmatically map unknown JSON data into the system. Given an incoming entity like:
```

Comment:
I think this is what we mean, since it is possible to map this as an `array` with no constraints on its `items` without introducing an `Empty List`. 
```suggestion
Similarly, the `Empty List` Data Type is required to be able to programmatically map unknown JSON data into the system without typing it as an unconstrained list. Given an incoming entity like:
```

Reply:
That's right. The general idea is that we don't want a plain "list" type because Update methods and such that could rely on types for validation might allow a user to expand the list and add anything they want, implicitly (and undesirably) adding some notion of an "any" type. By not allowing the list to have more elements added to it, it will be required to change the type to explicitly define what is inside the list.